subsys/bluetooth/gatt_dm.c

```diff
@@ -277,12 +295,29 @@ static void discovery_complete_error(struct bt_gatt_dm *dm, int err)
 	}
 }
 
+static struct gatt_discover_work_params {
+	struct k_work work;
+	struct bt_gatt_dm *dm;
+} gatt_work_data;
+
+static void gatt_discover_work(struct k_work *work)
+{
+	struct gatt_discover_work_params *wd =
+		CONTAINER_OF(work, struct gatt_discover_work_params, work);
+	__ASSERT(wd->dm->state_flags[STATE_ATTRS_LOCKED], "Attributes not locked");
```

Comment:
Also, I wonder if it is better to just return if `STATE_ATTRS_LOCKED` is not set than assert. Or of the `k_work_cancel` should also be called in `bt_gatt_dm_data_release`. E.g. if a user called `bt_gatt_dm_start`, then `discovery_process_attribute` or `discovery_process_service` callback was fired which submitted the work, but then user called `bt_gatt_dm_data_release` and `bt_gatt_dm_start` from a higher priority context. The chances for this to happen are very low since the work is submitted without delay, but better to protect against this as well.

Also, wonder if it should be:
```suggestion
	__ASSERT(atomic_test_bit(wd->dm->state_flags, STATE_ATTRS_LOCKED), "Attributes not locked");
```

Reply:
Changed to return, and added cancel to `bt_gatt_dm_data_release`